File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <set>
#include <cmath>

using namespace std;
string s;
set<int> sets;
vector<bool> used;

bool isPrime(int num) {
    if(num == 1) return false;
    for(int i=2; i<=sqrt(num); i++) {
        if(num%i == 0) return false;
    }
    return true;
}

void dfs(int cnt, int sum) {
    if(sum != 0) sets.insert(sum);
    if(cnt == s.size()) return;
    
    int tmp = -1;
    for(int i=0; i<s.size(); i++) {
        int curr = s[i]-'0';
        if(used[i] || tmp == curr) continue;
        
        used[i] = true;
        
        dfs(cnt+1, sum*10+curr);
        
        used[i] = false;
        tmp = curr;
    }
    
}

int solution(string numbers) {
    int answer = 0;
    s = numbers;
    sort(s.begin(), s.end());
    used.resize(s.size(), false);
    dfs(0,0);
    for(auto s : sets) {
        if(isPrime(s)) {
            answer++;
        }
    }
    return answer;
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    int answer = 0;
    int[] arr;
    boolean[] vis;
    HashSet<Integer> sets = new HashSet<>();
    
    boolean isPrime(int num) {
        if(num == 0 || num == 1) return false;
        for(int i=2; i<=Math.sqrt(num); i++) {
            if(num%i == 0) return false;
        }
        return true;
    }

    void dfs(int num) {
        sets.add(num);
        
        for(int i=0; i<arr.length; i++) {
            if(!vis[i]) {
                vis[i] = true;
                dfs(num*10 + arr[i]);
                vis[i] = false;
            }
        }
    }

    public int solution(String numbers) {
        arr = new int[numbers.length()];
        for(int i=0; i<numbers.length(); i++) {
            arr[i] = numbers.charAt(i) - '0';
        }
        vis = new boolean[arr.length];
        
        dfs(0);
        for(int s : sets) {
            // System.out.println(s);
            if(isPrime(s)) answer++;
        }
        
        return answer;
    }
}